subsys/testsuite/ztest/include/zephyr/ztest_assert.h

```diff
@@ -476,6 +549,149 @@ static inline bool z_zassume(bool cond, const char *default_msg, const char *fil
+ *
+ * @param a Value to compare
+ * @param b Value to compare
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_not_equal(a, b, ...) zexpect((a) != (b), #a " equal to " #b, ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a a equals @a b, otherwise mark test as failed but continue its execution.
+ *
+ * @a a and @a b will be converted to `void *` before comparing.
+ *
+ * @param a Value to compare
+ * @param b Value to compare
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_equal_ptr(a, b, ...)                                                               \
+	zexpect((void *)(a) == (void *)(b), #a " not equal to " #b, ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a a is within @a b with delta @a d, otherwise mark test as failed but
+ * continue its execution.
+ *
+ * @param a Value to compare
+ * @param b Value to compare
+ * @param d Delta
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_within(a, b, d, ...)                                                               \
+	zexpect(((a) >= ((b) - (d))) && ((a) <= ((b) + (d))), #a " not within " #b " +/- " #d,     \
+		##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a a is greater than or equal to @a l and less
+ *        than or equal to @a u, otherwise mark test as failed but continue its execution.
+ *
+ * @param a Value to compare
+ * @param l Lower limit
+ * @param u Upper limit
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_between_inclusive(a, l, u, ...)                                                    \
+	zexpect(((a) >= (l)) && ((a) <= (u)), #a " not between " #l " and " #u " inclusive",       \
+		##__VA_ARGS__)
+
+/**
+ * @brief Expect that 2 memory buffers have the same contents, otherwise mark test as failed but
+ * continue its execution.
+ *
+ * This macro calls the final memory comparison expectation macro.
+ * Using double expansion allows providing some arguments by macros that
+ * would expand to more than one values (ANSI-C99 defines that all the macro
+ * arguments have to be expanded before macro call).
+ *
+ * @param ... Arguments, see @ref zexpect_mem_equal__
+ *            for real arguments accepted.
+ */
+#define zexpect_mem_equal(...) zexpect_mem_equal__(__VA_ARGS__)
+
+/**
+ * @brief Internal expect that 2 memory buffers have the same contents, otherwise mark test as
+ * failed but continue its execution.
+ *
+ * @note This is internal macro, to be used as a second expansion.
+ *       See @ref zexpect_mem_equal.
+ *
+ * @param buf Buffer to compare
+ * @param exp Buffer with expected contents
+ * @param size Size of buffers
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_mem_equal__(buf, exp, size, ...)                                                   \
```

Comment:
Nit: If it's internal, should it be something like `z_zexpect_mem_equal__`?

Reply:
I see basically 2 usages of `zassert_mem_equal__` outside this file. I'm inclined to think they were mistakes, and we should try to avoid that kind of thing here.

Reply:
Agreed.

So is this comment just nonsense?

```
 * This macro calls the final memory comparison assertion macro.
 * Using double expansion allows providing some arguments by macros that
 * would expand to more than one values (ANSI-C99 defines that all the macro
 * arguments have to be expanded before macro call).
```

I'm not terribly sure I understand or agree with what it's saying.